### DISCUSSION TITLE
Elasticsearch: Support to show more columns than message and log level when visualizing logs in explore

### DISCUSSION THREAD

**Discussion by suikast42** [grafana/grafana#42319] [Observability]
Is it possible to add a config filed for Elasticsteach log fields so that the explore view show only that fields of an index ?

**Comment by marefr** [comment]
You can configure a message and a log level field in datasource settings, see https://grafana.com/docs/features/datasources/elasticsearch/#logs-beta for reference. Isn't this enough?

**Comment by Theoooooo** [comment]
But it would be pretty good if you can choose multiples fields to being displayed along the message_field_name option. This way, you can pre-configure explore web page with specific fields for better exploring purpose like this request feature i did #18316 instead of creating a dashboard with variables, tables ... (the ghetto way for me because the explore page is pretty good and compact for displaying lots of informations compare to tables.)

I also notice that, while setting the level_field_name, i'm still only getting gray color among the logs if the value is equal or greater than 3. It would be great if we can customise the color for each level_field_name value available in the indice like green between 1 and 3, orange for 4,5,6 and red for above.

**Comment by marefr** [comment]
@Theoooooo Re-labeled and changed title of this issue to reflect a feature request. 

Regarding level, it currently only support textual levels, see https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/logs.ts#L8-L24. Suggest opening a separate feature request for than one.

**Comment by ml-cms** [comment]
Also host, program and facility fields would be great to support syslogs from multiple sources.

**Comment by missyoyo** [comment]
> 
> 
> You can configure a message and a log level field in datasource settings, see https://grafana.com/docs/features/datasources/elasticsearch/#logs-beta for reference. Isn't this enough?

HI, I'm try use "Message field name" in datasource config,but it seems only support one filed,I try "Message field name=filed1" it seems works
 but when I try "Message field name=filed1,filed2,fild3",I only get null when explore

**Comment by strowi** [comment]
In general it would be very convenient to be able to configure the "message field name" in the query instead of the datasource.
E.g. having nginx-logs parsed and split into http-statuscode, geo, url, ... fields, there is no way to display multiple of these in one grafana panel (pls correct me if i'm wrong).

**Comment by detailyang** [comment]
@marefr message and level field are not enough sometimes. For example, the hostname field is always useful

**Comment by gabor** [comment]
to do this with elastic requires support for visualizing multiple fields in the logs visualization, so this is related: https://github.com/grafana/grafana/discussions/42315

also, another similar request, for cloudwatch: https://github.com/grafana/grafana/discussions/48776

**Comment by gabor** [comment]
in explore-mode, you can open log lines and use the eye icon to replace the logLine with the value of the field.
i understand that currently it's not possible to use the eye icon in panels, so perhaps that should be implemented (this is trickier, because i assume you want grafana to "remember" which fields you chose)

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82942. Feel free to continue the discussion around this there. Thank you!

**Comment by GitHamburg** [comment]
**In Log Panel**

Through the Transform data feature, the following can be achieved:

Add a new “Organize fields by name” operation:
1. Able to hide certain unnecessary fields
2. Can move the @message field to the first row

**In explore**
Toggle detected fields：

If your logs are structured in JSON or logfmt, you can show or hide the detected fields. Expand a log line and then **click the eye icon** to show or hide the fields.

**Comment by mabrarov** [comment]
Hi,

My team uses [`logsExploreTableVisualisation`](https://grafana.com/docs/grafana-cloud/whats-new/2023-12-13-logs-table-ui/) Grafana [feature toggle](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/) (where Grafana is deployed using kube-prometheus-stack Helm chart from https://github.com/prometheus-community/helm-charts/ and `logsExploreTableVisualisation` Grafana feature toggle is enabled using `GF_FEATURE_TOGGLES_ENABLE` environment variable):

> | Feature toggle name | Description | Enabled by default
> |---|---|---|
> | `logsExploreTableVisualisation` | A table visualisation for logs in Explore | Yes |

This Grafana feature toggle allows my team working with Grafana data source of Elasticsearch type in "Explore" Grafana mode to:

1. Chose visible columns
1. Chose order of columns

![Screenshot from 2024-08-19 22-16-22 (002)](https://github.com/user-attachments/assets/f59cf8db-2fe2-4849-8f29-163aea78defd)

Thank you.
